### ISSUE TITLE
System services are not being reschedule after power off

### ISSUE THREAD

**Issue by Dgotlieb** [hashicorp/nomad#12023]
### Nomad version
v1.1.2 (60638a086ef9630e2a9ba1e237e8426192a44244)

### Operating system and Environment details
Distributor ID:	Ubuntu
Description: Ubuntu 18.04.6 LTS
Release: 18.04
Codename: bionic

### Issue
After a power cut, `system` jobs are failing to restart and Nomad doesn't try to restart those jobs.

### Reproduction steps
While a `system` job is running, shutdown all cluster instances at once, then start them.

#### Expected Result
`system` job to be rescheduled.

#### Actual Result
`system` jobs are not rescheduled with a variety of errors, for example:
`failed to configure networking for alloc`
`Failed to start container XXXX: API error (404): No such container`

#### Workaround
Change job type to `service` and use [reschedule stanza](https://www.nomadproject.io/docs/job-specification/reschedule).

Due to `system` jobs nature they can't use the reschedule block, which forced me to change the job type to `service` and make sure to add the `count` as the number of Nomad clients while configuring a constraint so there are no 2 instances of this job on the same client:
```
    constraint {
      operator  = "distinct_hosts"
      value     = "true"
    } 
```

### Job file (before workaround):
```
job "foo" {
  type        = "system"
  datacenters = ["bar"]
  group "baz" {
    restart {
      attempts = 30
      delay    = "30s"
      mode     = "delay"
    }
    task "task" {
            driver = "docker"
            config {
            ...
      }
```

### Job file (before workaround):
```
job "foo" {
  type        = "service"
  datacenters = ["bar"]
  group "baz" {

    count = <THE_NUMBER_OF_NOMAD_CLIENTS>

    restart {
      attempts = 30
      delay    = "30s"
      mode     = "delay"
    }
    constraint {
      operator  = "distinct_hosts"
      value     = "true"
    }

    reschedule {
      unlimited = true
      delay_function = "constant"
      delay = "30s"
    }

    task "task" {
            driver = "docker"
            config {
            ...
      }
```

### Side notes
We are using Docker driver

**Comment by Dgotlieb** [comment]
Hi @tgross, thanks for the quick reply!

What you say makes total sense, but I wonder why the scheduler doesn't try to reschedule the failed `system` jobs..
I would expect the scheduler to treat `system` jobs as "high priority" jobs and try rescheduling them for unlimited amount of attempts (even in driver failures scenarios).
I know this can be a high resource consuming feature and may not always be necessary, so even a flag could be useful here.

1. The client nodes were gone for about 5 minutes.
This is my `server.hcl` config (if relevant):
```
    node_gc_threshold = "24h"
    eval_gc_threshold = "1h"
    job_gc_threshold = "4h"
    deployment_gc_threshold = "1h"
```

2. All services from type `service` (**not** `system`) got re-allocated.
Attached the `nomad alloc status` command output:

```
ID                  = b0db8302-e602-a266-f04d-edc0a9ab1940
Eval ID             = 7a3af95c
Name                = ingress.ingress[0]
Node ID             = 09355af8
Node Name           = nomad-server-2
Job ID              = ingress
Job Version         = 0
Client Status       = failed
Client Description  = Failed tasks
Desired Status      = run
Desired Description = <none>
Created             = 2h54m ago
Modified            = 1m31s ago

Allocation Addresses (mode = "bridge")
Label                                 Dynamic  Address
*AAA             yes      10.10.0.113:5555 -> 5555

Task "ingress-gateway-poststart" (poststart) is "dead"
Task Resources
CPU      Memory   Disk     Addresses
100 MHz  300 MiB  300 MiB  

Task Events:
Started At     = 2022-02-09T08:55:33Z
Finished At    = 2022-02-09T08:55:34Z
Total Restarts = 0
Last Restart   = N/A

Recent Events:
Time                       Type           Description
2022-02-09T13:48:31+02:00  Setup Failure  failed to setup alloc: pre-run hook "network" failed: failed to configure networking for alloc: failed to configure network: failed to allocate for range 0: 172.26.70.193 has been allocated to b0db8302-e602-a266-f04d-edc0a9ab1940, duplicate allocation is not allowed
2022-02-09T13:48:18+02:00  Killing        Sent interrupt. Waiting 5s before force killing
2022-02-09T10:55:34+02:00  Terminated     Exit Code: 0
2022-02-09T10:55:33+02:00  Started        Task started by client
2022-02-09T10:55:24+02:00  Task Setup     Building Task Directory
2022-02-09T10:55:22+02:00  Received       Task received by client

Task "connect-ingress-ingress" is "dead"
Task Resources
CPU         Memory       Disk     Addresses
0/3000 MHz  0 B/1.0 GiB  300 MiB  

Task Events:
Started At     = 2022-02-09T08:55:24Z
Finished At    = 2022-02-09T11:48:31Z
Total Restarts = 0
Last Restart   = N/A

Recent Events:
Time                       Type           Description
2022-02-09T13:48:31+02:00  Setup Failure  failed to setup alloc: pre-run hook "network" failed: failed to configure networking for alloc: failed to configure network: failed to allocate for range 0: 172.26.70.193 has been allocated to b0db8302-e602-a266-f04d-edc0a9ab1940, duplicate allocation is not allowed
2022-02-09T13:48:18+02:00  Killing        Sent interrupt. Waiting 5s before force killing
2022-02-09T10:55:24+02:00  Started        Task started by client
2022-02-09T10:55:24+02:00  Task Setup     Building Task Directory
2022-02-09T10:55:22+02:00  Received       Task received by client
```

It may be worth mentioning that this is happening also when running the job without the `poststart` task. 


Thanks!

**Comment by Dgotlieb** [comment]
@tgross 
The second option:

> Or was this for an allocation that was running before the poweroff and now got this Setup Failure event when the client tried to come back?

**Comment by tgross** [comment]
Ok that's interesting. I think in this case we're in a state where the allocation _can't_ be recovered (which may be a bug in and of itself!) no matter how many times the scheduler tries to place it. `nomad job status -verbose ingress` may have more details about the scheduler restart decisions, but also it might be helpful for me to figure out the networking problem if you could provide the job spec for the `ingress` job as well.

**Comment by Dgotlieb** [comment]
@tgross, I saw https://github.com/hashicorp/nomad/issues/11272 and the networking issue is probably related.
That being said, my issue is not only related to the networking issue and has multiple errors including: 
`Failed to start container XXXX: API error (404): No such container`.

As I wrote before my assumption is that when the `system` type is in use, Nomad "gives up" on the first driver failure and doesn't try to recover from it by waiting for the Docker driver / CNI plugin to be healthy again (contrary to when using the `reschedule stanza` on a `service` type).  

Attaching the job spec:
```
job "ingress" {
  type        = "system"
  datacenters = ["dc1"]
  group "ingress" {
    restart {
      attempts = 30
      delay    = "30s"
      mode     = "delay"
    }

    network {
      mode = "bridge"
      port "aaa" {
        static = 5555
        to     = 5555
      }

      port "bbb" {
        static = 8080
        to     = 8080
      }

      port "ccc" {
        static = 8000
        to     = 8000
      }
      ...
    }

    service {
      name = "ingress"
      connect {
        sidecar_task {
          resources {
            cpu    = 10000
            memory = 1024
          }
        }

        gateway {
          proxy {}
          ingress {
            listener {
              port = 5555
              service {
                name = "aaa"
              }
            }

            listener {
              port = 8080
              service {
                name = "bbb"
              }
            }
            
            listener {
              port = 8000
              service {
                name = "ccc"
              }
            }
            ...
          }
        }
      }
    }

    task "ingress-gateway-poststart" {
      lifecycle {
        hook    = "poststart"
        sidecar = false
      }

      template {
        data        = file("ingress_poststart.py")
        destination = "local/ingress_poststart.py"
      }

      driver = "exec"
      config {
        command = "python3"
        args    = ["local/ingress_poststart.py"]
      }
    }
  }
}
```


And the status output (`$nomad job status -verbose ingress`):
```
ID            = ingress
Name          = ingress
Submit Date   = 2022-02-10T12:48:49+02:00
Type          = system
Priority      = 50
Datacenters   = dc1
Namespace     = default
Status        = running
Periodic      = false
Parameterized = false

Summary
Task Group  Queued  Starting  Running  Failed  Complete  Lost
ingress     0       0         0        3       0         0

Evaluations
ID                                    Priority  Triggered By    Status    Placement Failures
344e7ced-9faf-d575-4f93-57d32e1f6443  50        node-update     complete  false
8c20f3ce-e9c7-7bd6-822c-0996974b6eac  50        node-update     complete  false
4ce7672e-1828-ffc3-0420-60abd9a80f69  50        node-update     complete  false
b97a3855-ab74-8c5a-4a14-bd81c988770e  50        job-register    complete  false
b3f7e2b7-e273-eec0-7e23-85e8b6c68eac  50        job-deregister  complete  false
7a3af95c-1e2c-7c1e-0511-9bdbf95ae6a8  50        job-register    complete  false

Allocations
ID                                    Eval ID                               Node ID                               Node Name       Task Group  Version  Desired  Status  Created                    Modified
32bb6c62-ea34-8667-91ce-d777d15c7228  b97a3855-ab74-8c5a-4a14-bd81c988770e  a9f5963f-e513-9f40-94f8-d760d94b48b7  nomad-server-1  ingress     0        run      failed  2022-02-10T12:48:49+02:00  2022-02-10T12:52:25+02:00
47b014fa-d4f1-3416-98af-3ecf6f8301dc  b97a3855-ab74-8c5a-4a14-bd81c988770e  584a9dc7-a59d-ec14-505f-dc3bcdc18c88  nomad-server-3  ingress     0        run      failed  2022-02-10T12:48:49+02:00  2022-02-10T12:52:29+02:00
712353fb-da69-5a1b-fc55-016fb0971088  b97a3855-ab74-8c5a-4a14-bd81c988770e  09355af8-370a-0a05-3a79-027fc5d384e5  nomad-server-2  ingress     0        run      failed  2022-02-10T12:48:49+02:00  2022-02-10T12:52:26+02:00

```

**Comment by Dgotlieb** [comment]
@tgross, I'm sure this issue will require further investigation, but I just wanted to ask if you have any insights on my workaround (using `service` type with `reschedule` + `distinct_hosts` constraint) or if there are caveats that should be considered.

Thanks

**Comment by tgross** [comment]
I think I missed previously that this was a `poststart` task, but I see you mentioned earlier:

> It may be worth mentioning that this is happening also when running the job without the poststart task.

Is "this is happening" referring to the network error or not rescheduling the allocation? I would expect the semantics around sidecar/poststart tasks to be different than the primary task, and it's not clear whether the primary task is failing, whether it's the alloc as a whole, etc.

Can we get the _entire_ jobspec (redacted if necessary) and the entire `alloc status`? I think not having that is making it harder to get my head around what's happening here because I'm not seeing the alloc status output from the primary task.

**Comment by Dgotlieb** [comment]
@tgross, **"this is happening"**, refers to the failure of rescheduling an allocation after the reboot.

The job spec was already provided in my previous comment and given that this is an ingress job, we **don't** have a main task (only a `service` and a `poststart` task).

That being said, it is important to mention that the issue is also happening on the below 2 options as well, so I'm not sure this is related to the lifecycle hook of the task (`main`/`poststart`) but of course I could be wrong:
1. `system` service **without** a `poststart` task
2. `system` service **with** main task, but **without** a `poststart` task

**Comment by tgross** [comment]
Ok @Dgotlieb I spent some time digging into this and trying a bunch of different ways to replicate this behavior. What it comes down is that `reschedule` behavior doesn't apply to `system` jobs. So when the job exceeds the `restart` attempts, it's forever dead until external state changes and kicks off an evaluation for the job.

You've (correctly) avoided that scenario by using `restart.mode = "delay"`, which only makes the task wait a while before restarting. But with some experiments with an intentionally broken client environment, I was able to reproduce when I hit the following error:

> 2022-03-14T18:44:44.845Z [WARN]  client.alloc_runner.runner_hook: failed to configure network: alloc_id=c0b753e2-f832-a8f2-9995-0a5545f453b5 err="plugin type=\"bridge\" failed (add): failed to set bridge addr: could not set bridge's mac: invalid argument" attempt=1

I suspect what's happening here is that when the client can't complete the hooks that happen before the restart count is set up, then we can fail the whole allocation. And because there's no reschedule evaluation made, we never get a running allocation until a new evaluation for the job is run.

My takeaways on these are two:
* I've kicked off a discussion internally about whether not firing a new evaluation on [`(*Node) UpdateAlloc()`](https://github.com/hashicorp/nomad/blob/v1.2.6/nomad/node_endpoint.go#L1150) is actually the right behavior. It smells wrong to me, but it's entirely possible I'm missing something.
* We should probably add a note to the docs clarifying this behavior.

**Comment by tgross** [comment]
I've marked this issue for roadmapping to give it some more attention.

**Comment by Dgotlieb** [comment]
@tgross thanks for researching and roadmapping this issue!
Looking forward for the fix 😅

**Comment by Rendanic** [comment]
What is the state of this issue? Is it fixed or any documentation how to fix it with proper job configuration?

Regards
Thorsten

**Comment by mwild1** [comment]
I can easily reproduce this behaviour by simply restarting the docker daemon on a client node. Allocations of a system job on that node then say "desired: run, status: failed" but nomad never attempts to restart/reschedule them.

The whole reason I'm using Nomad for this is because I want this container running on every box at all times. Is there a better way to achieve that, or is it just not possible right now?

Note that I've only tested up to Nomad 1.4 so far, but I'd be happy to spend time upgrading if there's a chance this has been fixed (I currently see no indication that it has). I'm also happy to pull any useful debugging info, or even grant Nomad team members temporary access to the systems (it's a dev environment, though this issue has bitten me in production multiple times in the past few years).

**Comment by dcarbone** [comment]
This also causes issues if a `template` produces an error in the job definition.  The job perma-fails and is never rescheduled.

**Comment by axxapy** [comment]
We observe same behavior with nomad 1.10.4

**Comment by superboum** [comment]
We also hit this issue sometimes, especially when our network is degraded (with a high ratio of packet loss). Nomad tends to fall in a broken state where all our system services are dead despite having a proper `restart` block set with `delay`. During this degraded network event, we probably encounter an error similar to what @tgross found, a kind of error that appears before the `restart` hook is evaluated.

**TL;DR:** The contribution of this comment is: degraded network performances, especially high packet loss, can lead to this situation.
